metricflow/plan_conversion/dataflow_to_sql.py

```diff
@@ -1331,35 +1341,48 @@ def visit_join_to_time_spine_node(self, node: JoinToTimeSpineNode) -> SqlDataSet
             SqlColumnReference(table_alias=time_spine_alias, column_name=time_spine_dim_instance.spec.qualified_name)
         )
 
-        # Add requested granularities (if different from time_spine) and date_parts to time spine column.
         time_spine_select_columns = []
         time_spine_dim_instances = []
-        where: Optional[SqlExpressionNode] = None
-        for requested_time_dimension_spec in node.requested_agg_time_dimension_specs:
-            # Apply granularity to time spine column select expression.
-            if requested_time_dimension_spec.time_granularity == time_spine_dim_instance.spec.time_granularity:
-                select_expr: SqlExpressionNode = time_spine_column_select_expr
-            else:
-                select_expr = SqlDateTruncExpression(
-                    time_granularity=requested_time_dimension_spec.time_granularity, arg=time_spine_column_select_expr
+        where_filter: Optional[SqlExpressionNode] = None
+
+        # If offset_to_grain is used, will need to filter down to rows that match selected granularities.
+        # Does not apply if one of the granularities selected matches the time spine column granularity.
+        need_where_filter = (
+            node.offset_to_grain and time_spine_dim_instance.spec not in node.requested_agg_time_dimension_specs
+        )
+
+        # Add requested granularities (if different from time_spine) and date_parts to time spine column.
+        for time_dimension_instance in time_dimensions_to_select_from_time_spine:
+            time_dimension_spec = time_dimension_instance.spec
+
+            # Apply grain to time spine column select expression, unless grain already matches time spine column.
+            select_expr: SqlExpressionNode = (
+                time_spine_column_select_expr
+                if time_dimension_spec.time_granularity == time_spine_dim_instance.spec.time_granularity
+                else SqlDateTruncExpression(
+                    time_granularity=time_dimension_spec.time_granularity, arg=time_spine_column_select_expr
```

Comment:
Good point! I feel like the time spine SHOULD be set with the smallest possible grain in order to fill all granularity periods we support.
In practice, though, it might not be - especially when we transition to supporting smaller grains. Users might need to update their time spines.
I'll add an assertion for now and a comment to explain!